#!/applications/R/R-3.5.0/bin/Rscript

# Convert samtools pileup file into a haplotype matrix
# for visualisation of recombination over interval

#sample <- "barcode1_120"
#MAPQ <- 60

args <- commandArgs(trailingOnly = T)
sample <- args[1]
MAPQ <- args[2]

setwd("/home/ajt200/analysis/nanopore/pollen_typing/")
library(dplyr)
library(parallel)
library(doParallel)
registerDoParallel(cores = detectCores())
print("Currently registered parallel backend name, version and cores")
print(getDoParName())
print(getDoParVersion())
print(getDoParWorkers())

# Load reference and alternate alleles
alleles <- read.table("3a_SNPs_in_Col_and_Ws.tsv",
                      header = T)

# Load pileup containing per-alignment genotype information,
# as generated by samtools mpileup
plp <- read.table(paste0("minimap2_TAIR10_",
                         sample,
                         "_sorted_primary_MAPQ", MAPQ, ".plp"),
                  header = F, sep = "\t",
                  quote = "", comment.char = "",
                  stringsAsFactors = F)

# Append the first 3 columns, the alternate allele, and columns
# corresponding to per-alignment file genotype
plp <- data.frame(plp[ ,1:3],
                  alleles[ ,4],
                  plp[ ,seq(from = 5, 
                            to = dim(plp)[2]-1,
                            by = 3)])
colnames(plp) <- c("chr", "pos", "ref", "alt",
                   seq(1:(dim(plp)[2]-4)))
# Replace factors with characters
factorColumns <- sapply(plp, is.factor)
plp[factorColumns] <- lapply(plp[factorColumns], as.character)
plpHap <- plp

# Re-encode haplotypes using genotype naming convention
# to avoid incorrect encoding of adenosine bases
# Col-0 as "AA", Ws-4 as "BB", "*" and all others as "-" 
for(x in 1:(dim(plpHap)[1])) {
  for(y in 5:(dim(plpHap)[2])) {
    if( plpHap[x,y] %in% c(".", ",") ) {
      plpHap[x,y] <- "AA"
    } else if( plpHap[x,y] %in% c(plpHap[x,4],
                                  tolower(plpHap[x,4])) ) {
      plpHap[x,y] <- "BB"
    } else {
      plpHap[x,y] <- "-"
    }
  }
}
# Convert genotype naming convention into
# haplotype naming convention as we cannot derive
# each allele from single-molecule sequencing
plpHap[] <- lapply(plpHap, function(x) as.character(gsub("AA", "A", x)))
plpHap[] <- lapply(plpHap, function(x) as.character(gsub("BB", "B", x)))

# Transpose haplotype matrix for sorting
tplpHap <- data.frame(t(plpHap[,5:dim(plpHap)[2]]))
colnames(tplpHap) <- as.integer(plpHap$pos)

# Add column representing haplotypes as character strings
stringHap <- sapply(1:(dim(tplpHap)[1]), function(x) {
  paste0(as.character(as.matrix(tplpHap[x,])), collapse = "")
})
tplpHap <- data.frame(tplpHap,
                      hap = stringHap)
colnames(tplpHap) <- c(as.integer(plpHap$pos),
                       "hap")
# Replace factors with characters
factorColumns <- sapply(tplpHap, is.factor)
tplpHap[factorColumns] <- lapply(tplpHap[factorColumns], as.character)

# Retain only rows that contain both parental haplotypes
tplpHap <- tplpHap[grepl("A", tplpHap$hap) &
                   grepl("B", tplpHap$hap),]

# Apply a minimal haplotype imputation approach which
# is informed only by haplotypes at flanking markers
# and not by haplotypes within other alignments
# For Col-0 (A) alleles
while( sum(grepl(pattern = "A(-+)A",
                 x = tplpHap$hap,
                 perl = T)) > 0 ) {
  for(x in 1:length(tplpHap$hap)) {
    if( grepl(pattern = "A(-+)A",
              x = tplpHap$hap[x],
              perl = T) ) {
      tplpHap$hap[x] <- sub(pattern = "A(-+)A",
                            replacement = paste0("A",
                                                 paste0(rep(x = "A",
                                                            times = attr(regexpr(pattern = "A(-+)A",
                                                                                 text = tplpHap$hap[x],
                                                                                 perl = T),
                                                                         'capture.length')[1]),
                                                        collapse = ""),
                                                 "A"),
                            x = tplpHap$hap[x],
                            perl = T)
    }
  }
}
# For Ws-4 (B) alleles
while( sum(grepl(pattern = "B(-+)B",
                 x = tplpHap$hap,
                 perl = T)) > 0 ) {
  for(x in 1:length(tplpHap$hap)) {
    if( grepl(pattern = "B(-+)B",
              x = tplpHap$hap[x],
              perl = T) ) {
      tplpHap$hap[x] <- sub(pattern = "B(-+)B",
                            replacement = paste0("B",
                                                 paste0(rep(x = "B",
                                                            times = attr(regexpr(pattern = "B(-+)B",
                                                                                 text = tplpHap$hap[x],
                                                                                 perl = T),
                                                                         'capture.length')[1]),
                                                        collapse = ""),
                                                 "B"),
                            x = tplpHap$hap[x],
                            perl = T)
    }
  }
}

# Impute haplotypes at leftmost and rightmost markers based
# on haplotype at nearest markers
# For Col-0 (A) alleles
# leftmost
while( sum(grepl(pattern = "^(-+)AA",
                 x = tplpHap$hap,
                 perl = T)) > 0 ) {
  for(x in 1:length(tplpHap$hap)) {
    if( grepl(pattern = "^(-+)AA",
              x = tplpHap$hap[x],
              perl = T) ) {
      tplpHap$hap[x] <- sub(pattern = "^(-+)AA",
                            replacement = paste0(paste0(rep(x = "A",
                                                            times = attr(regexpr(pattern = "^(-+)AA",
                                                                                 text = tplpHap$hap[x],
                                                                                 perl = T),
                                                                         'capture.length')[1]),
                                                        collapse = ""),
                                                 "AA"),
                            x = tplpHap$hap[x],
                            perl = T)
    }
  }
}
# rightmost
while( sum(grepl(pattern = "AA(-+)$",
                 x = tplpHap$hap,
                 perl = T)) > 0 ) {
  for(x in 1:length(tplpHap$hap)) {
    if( grepl(pattern = "AA(-+)$",
              x = tplpHap$hap[x],
              perl = T) ) {
      tplpHap$hap[x] <- sub(pattern = "AA(-+)$",
                            replacement = paste0("AA",
                                                 paste0(rep(x = "A",
                                                            times = attr(regexpr(pattern = "AA(-+)$",
                                                                                 text = tplpHap$hap[x],
                                                                                 perl = T),
                                                                         'capture.length')[1]),
                                                        collapse = "")),
                            x = tplpHap$hap[x],
                            perl = T)
    }
  }
}

# For Ws-4 (B) alleles
# leftmost
while( sum(grepl(pattern = "^(-+)BB",
                 x = tplpHap$hap,
                 perl = T)) > 0 ) {
  for(x in 1:length(tplpHap$hap)) {
    if( grepl(pattern = "^(-+)BB",
              x = tplpHap$hap[x],
              perl = T) ) {
      tplpHap$hap[x] <- sub(pattern = "^(-+)BB",
                            replacement = paste0(paste0(rep(x = "B",
                                                            times = attr(regexpr(pattern = "^(-+)BB",
                                                                                 text = tplpHap$hap[x],
                                                                                 perl = T),
                                                                         'capture.length')[1]),
                                                        collapse = ""),
                                                 "BB"),
                            x = tplpHap$hap[x],
                            perl = T)
    }
  }
}
# rightmost
while( sum(grepl(pattern = "BB(-+)$",
                 x = tplpHap$hap,
                 perl = T)) > 0 ) {
  for(x in 1:length(tplpHap$hap)) {
    if( grepl(pattern = "BB(-+)$",
              x = tplpHap$hap[x],
              perl = T) ) {
      tplpHap$hap[x] <- sub(pattern = "BB(-+)$",
                            replacement = paste0("BB",
                                                 paste0(rep(x = "B",
                                                            times = attr(regexpr(pattern = "BB(-+)$",
                                                                                 text = tplpHap$hap[x],
                                                                                 perl = T),
                                                                         'capture.length')[1]),
                                                        collapse = "")),
                            x = tplpHap$hap[x],
                            perl = T)
    }
  }
}


# Order by (imputed) haplotypes
tplpHap_sort <- tplpHap[sort.int(tplpHap$hap,
                                 decreasing = T,
                                 index.return=T)$ix,]
# Group by (imputed) haplotypes
tplpHap_group_n <- tplpHap %>%
  group_by(hap) %>%
  summarize(n())


tmp <- arrange(tplpHap,
               desc(tplpHap[,1]),
	       desc(tplpHap[,2]),
               desc(tplpHap[,3]),
               desc(tplpHap[,4]),
               desc(tplpHap[,5]),
               desc(tplpHap[,6]),
               desc(tplpHap[,7]),
               desc(tplpHap[,8]),
               desc(tplpHap[,9]),
               desc(tplpHap[,10]),
               desc(tplpHap[,11]),
               desc(tplpHap[,12]),
               desc(tplpHap[,13]),
               desc(tplpHap[,14]),
               desc(tplpHap[,15]),
               desc(tplpHap[,16]))

tmp2 <- arrange(tbl_df(tplpHap),
               desc(tplpHap[,1:16]))
	       desc(tplpHap[,2]),
               desc(tplpHap[,3]),
               desc(tplpHap[,4]),
               desc(tplpHap[,6]),
               desc(tplpHap[,6]),
               desc(tplpHap[,7]),
               desc(tplpHap[,8]),
               desc(tplpHap[,9]),
               desc(tplpHap[,10]),
               desc(tplpHap[,11]),
               desc(tplpHap[,12]),
               desc(tplpHap[,13]),
               desc(tplpHap[,14]),
               desc(tplpHap[,15]),
               desc(tplpHap[,16]))

tmp <- tplpHap[order(
                     tplpHap[,1],
		     tplpHap[,2],
                     tplpHap[,3],
                     tplpHap[,4],
                     tplpHap[,6],
                     tplpHap[,6],
                     tplpHap[,7],
                     tplpHap[,8],
                     tplpHap[,9],
                     tplpHap[,10],
                     tplpHap[,11],
                     tplpHap[,12],
                     tplpHap[,13],
                     tplpHap[,14],
                     tplpHap[,15],
                     tplpHap[,16]
                    ),]
