#!/applications/R/R-3.6.0/bin/Rscript

# Convert samtools pileup file into a haplotype matrix
# for visualisation of recombination over interval

# Usage:
# ./pileup_to_haplotype_load_allelesMat.R barcode1_120 60 0.153  

#sample <- "barcode1_120"
#MAPQ <- 60
#cMscale <- 0.153

args <- commandArgs(trailingOnly = T)
sample <- args[1]
MAPQ <- args[2]
cMscale <- as.numeric(args[3])

setwd("/home/ajt200/analysis/nanopore/pollen_typing/")
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggthemes)
library(grid)
library(gridExtra)
library(extrafont)
font_import()
loadfonts()
library(ComplexHeatmap)
library(Cairo)
library(png)
library(parallel)
library(doParallel)
registerDoParallel(cores = detectCores())
print("Currently registered parallel backend name, version and cores")
print(getDoParName())
print(getDoParVersion())
print(getDoParWorkers())

hapMatDir <- "haplotype_matrices/"
plotDir <- "plots/"
system(paste0("[ -d ", hapMatDir, " ] || mkdir ", hapMatDir))
system(paste0("[ -d ", plotDir, " ] || mkdir ", plotDir))

# Load reference and alternate alleles
alleles <- read.table("3a_SNPs_indels_in_Col_and_Ws.tsv",
                      header = T)
# Replace factors with characters
factorColumns <- sapply(alleles, is.factor)
alleles[factorColumns] <- lapply(alleles[factorColumns], as.character)
# Re-specify Ws-4 deletion allele at Chr3:637898 in row 8 of "alleles"
## NOTE: This could be made more or less specific to the known Col allele "ATGTTT" (vs. "AAAAA" in Ws-4)
## to capture sequences lacking either complete or partial matches to "TGTTT"
alleles[8, ]$Ws.4 <- "[.,]-[1-9]TGT"
# Re-specify Ws-4 alleles at Chr3:638788 in "alleles"
## NOTE: In the following two cases of insertions in Ws-4, the alternate allele specified
## could be made more specific to the known Ws-4 allele "TAT" (vs. "T" in Col) to capture
## sequences containing exact matches to the "AT" insertion in Ws-4
## However, only putative insertions with additional inserted bases (e.g., "ATA") were
## observed in the 5 samples
alleles[20, ]$Ws.4 <- "[.,]\\+[1-9]AT"
# Re-specify Ws-4 alleles at Chr3:639539 in "alleles"
alleles[23, ]$Ws.4 <- "[.,]\\+[1-9]AT"
# The alternate alleles for rows 20 and 23 (Chr3:638788 and 639539) appear not to be
# correctly called using ONT reads, possibly due the limited ability of this technology to generate
# sequences of repeat arrays that represent the true number of repeats in an array
# We observe largely ostensible reference alleles at these markers (e.g., 44163/55803 [79%] and 48451/55803 [87%]),
# which is improbable given that these alignments represent recombinant molecules which
# have the alternate (Ws-4) alleles at the rightmost marker
# (located within the Ws-4 allele-specific primer binding site)
# Therefore, remove rows 20 and 23 (Chr3:638788 and 639539)
alleles <- alleles[c(-20, -23), ]
# Remove rows 9, 10, 11, 12 and 13 (Chr3:637899, 637900, 637901, 637902, and 637903),
# (use information in row 8 (Chr3:637898) to obtain genotype at this indel polymorphism;
# ATGTTT in Col, AAAAA in Ws-4)
alleles <- alleles[-9:-13, ]
rownames(alleles) <- 1:dim(alleles)[1]

# Load pileup containing per-alignment genotype information,
# as generated by samtools mpileup
#plp <- read.table(paste0("minimap2_TAIR10_",
#                         sample,
#                         "_sorted_primary_MAPQ", MAPQ, ".plp"),
#                  header = F, sep = "\t",
#                  quote = "", comment.char = "",
#                  stringsAsFactors = F)
## Remove rows 20 and 23 (Chr3:638788 and 639539)
## Remove rows 9, 10, 11, 12 and 13 (Chr3:637899, 637900, 637901, 637902, and 637903),
## (use information in row 8 (Chr3:637898) to obtain genotype at this indel polymorphism;
## ATGTTT in Col, AAAAA in Ws-4)
#plp <- plp[c(-20, -23), ]
#plp <- plp[-9:-13, ]
#rownames(plp) <- 1:dim(plp)[1]
#
## Append the first 3 columns, the alternate allele, and columns
## corresponding to per-alignment file genotype
#plp <- data.frame(plp[ ,1:3],
#                  alleles[ ,4],
#                  plp[ ,seq(from = 5, 
#                            to = dim(plp)[2]-1,
#                            by = 3)])
#colnames(plp) <- c("chr", "pos", "ref", "alt",
#                   seq(1:(dim(plp)[2]-4)))
## Replace factors with characters
#factorColumns <- sapply(plp, is.factor)
#plp[factorColumns] <- lapply(plp[factorColumns], as.character)
#write.table(plp,
#            file = paste0(hapMatDir, sample, "_ONT_pileup_alleles.tsv"),
#            quote = F, sep = "\t",
#            row.names = F, col.names = T)
#plpHap <- plp

## Re-encode haplotypes using genotype naming convention
## to avoid incorrect encoding of adenosine bases
## Col-0 as "AA", Ws-4 as "BB", "*" and all others as "-" 
#for(x in 1:(dim(plpHap)[1])) {
#  for(y in 5:(dim(plpHap)[2])) {
#    print(paste0("Marker ", x, " of ", dim(plpHap)[1],
#                 ": Alignment ", y-4, " of ", dim(plpHap)[2]-4))
#    if( plpHap[x,y] %in% c(".", ",", "^].", "^],", ".$", ",$") ) {
#      plpHap[x,y] <- "AA"
#    } else if( plpHap[x,y] %in% c(plpHap[x,4],
#                                  tolower(plpHap[x,4]),
#                                  paste0("^]", plpHap[x,4]),
#                                  paste0("^]", tolower(plpHap[x,4])), 
#                                  paste0(plpHap[x,4], "$"),
#                                  paste0(tolower(plpHap[x,4]), "$")) ) {
#      plpHap[x,y] <- "BB"
#    } else if( grepl(pattern = plpHap[x,4],
#                     x = plpHap[x,y],
#                     ignore.case = T,
#                     perl = T) ) {
#      plpHap[x,y] <- "BB"
#    } else {
#      plpHap[x,y] <- "-"
#    }
#  }
#}
## Convert genotype naming convention into
## haplotype naming convention as we cannot derive
## each allele from single-molecule sequencing
#plpHap[] <- lapply(plpHap, function(x) as.character(gsub("AA", "A", x)))
#plpHap[] <- lapply(plpHap, function(x) as.character(gsub("BB", "B", x)))
#
#write.table(plpHap,
#            file = paste0(hapMatDir, sample, "_ONT_pileup_to_haplotype.tsv"),
#            quote = F, sep = "\t",
#            row.names = F, col.names = T)

plp2 <- read.table(paste0(hapMatDir, sample, "_ONT_pileup_alleles.tsv"),
                   header = T, sep = "\t",
                   stringsAsFactors = F)
colnames(plp2) <- c("chr", "pos", "ref", "alt",
                    seq(1:(dim(plp2)[2]-4)))

# Re-encode haplotypes using genotype naming convention
# to avoid incorrect encoding of adenosine bases
# Col-0 as "AA", Ws-4 as "BB", "*" and all others as "-" 
for(x in 1:(dim(plp2)[1])) {
  for(y in 5:(dim(plp2)[2])) {
    print(paste0("Marker ", x, " of ", dim(plp2)[1],
                 ": Alignment ", y-4, " of ", dim(plp2)[2]-4))
    if( plp2[x,y] %in% c(".", ",", "^].", "^],", ".$", ",$") ) {
      plp2[x,y] <- "AA"
    } else if( plp2[x,y] %in% c(plp2[x,4],
                                tolower(plp2[x,4]),
                                paste0("^]", plp2[x,4]),
                                paste0("^]", tolower(plp2[x,4])), 
                                paste0(plp2[x,4], "$"),
                                paste0(tolower(plp2[x,4]), "$")) ) {
      plp2[x,y] <- "BB"
    } else if( grepl(pattern = plp2[x,4],
                     x = plp2[x,y],
                     ignore.case = T,
                     perl = T) ) {
      plp2[x,y] <- "BB"
    } else if( plp2[x,y] %in% c("A", "C", "G", "T",
                                "a", "c", "g", "t",
                                "^]A", "^]C", "^]G", "^]T",
                                "^]a", "^]c", "^]g", "^]t",
                                "A$", "C$", "G$", "T$",
                                "a$", "c$", "g$", "t$") ) {
      plp2[x,y] <- "XX"
    }
  }
}
write.table(plp2,
            file = paste0(hapMatDir, sample, "_ONT_pileup_alleles_errors.tsv"),
            quote = F, sep = "\t",
            row.names = F, col.names = T)

af <- data.frame()
for(x in 1:dim(plp2)[1]) {
  marker_total <- length( plp2[x,5:dim(plp2)[2]] )
  marker_A <- sum( ( plp2[x,5:dim(plp2)[2]] %in% c("AA") ) )
  marker_B <- sum( ( plp2[x,5:dim(plp2)[2]] %in% c("BB") ) )
  marker_X <- sum( ( plp2[x,5:dim(plp2)[2]] %in% c("XX") ) )
  marker_N <- sum( ( plp2[x,5:dim(plp2)[2]] %in% c("*") ) )
  marker_I <- sum(!( plp2[x,5:dim(plp2)[2]] %in% c("AA", "BB", "XX", "*") ) )
  stopifnot(marker_total ==
            sum(marker_A, marker_B, marker_X, marker_N, marker_I))
  af_x <- data.frame(marker = alleles$pos[x],
                     A = marker_A/marker_total,
                     B = marker_B/marker_total,
                     X = marker_X/marker_total,
                     I = marker_I/marker_total,
                     N = marker_N/marker_total)
  af <- rbind(af, af_x)
}

# Remove markers within allele-specific primer binding sites
af <- af[c(-1, -dim(af)[1]), ]
af_tidy <- gather(data = af,
                  key = allele,
                  value = af,
                  -marker)
af_tidy$allele <- factor(af_tidy$allele,
                         levels = unique(af_tidy$allele))

# Define legend labels
alleleNames <- c("Col-0", "Ws-4", "Nonparental SNV", "Nonparental indel", "Missing")
alleleColours <- c("red", "blue", "darkorange2", "green2", "grey40") 
makeTransparent <- function(thisColour, alpha = 180)
{
  newColour <- col2rgb(thisColour)
  apply(newColour, 2, function(x) {
    rgb(red = x[1], green = x[2], blue = x[3],
        alpha = alpha, maxColorValue = 300)
  })
}
#alleleColours <- makeTransparent(alleleColours)
legendPos <- as.numeric(unlist(strsplit("0.01,0.96",
                                        split = ",")))
legendLabs <- lapply(seq_along(alleleNames), function(x) {
  grobTree(textGrob(bquote(.(alleleNames[x])),
                    x = legendPos[1], y = legendPos[2]-((x-1)*0.06), just = "left",
                    gp = gpar(col = alleleColours[x], fontsize = 18)))
})

# Plot allele frequencies for each marker
ggObj_af <- ggplot(data = af_tidy,
                   mapping = aes(x = marker,
                                 y = af,
                                 fill = allele)) +
   geom_bar(position = "fill", stat = "identity", width = 14) +
#  geom_segment(mapping = aes(x = marker, y = 0,
#                             xend = marker, yend = af,
#                             colour = allele),
#               size = 1.5) +
  scale_fill_manual(values = alleleColours) +
  scale_x_continuous(breaks = c(alleles$position[2:(length(alleles$position)-1)]),
                     labels = c(as.character(alleles$position[2:(length(alleles$position)-1)]))) +
  scale_y_continuous(limits = c(-0.1, 1.0),
                     breaks = c(0.0, 0.2, 0.4, 0.6, 0.8, 1.0),
                     labels = function(x) sprintf("%1.1f", as.numeric(x))) +
  geom_vline(xintercept = c(alleles$position[2:(length(alleles$position)-1)]),
             linetype = "dashed",
             size = 0.5) +
  # Add genes within 3a hotspot
  # 3       634653  637284  +       AT3G02880.1
  # 3       638021  639055  -       AT3G02885.1
  geom_segment(mapping = aes(x = 634653, y = -0.07,
                             xend = 637284, yend = -0.07),
               arrow = arrow(angle = 10, type = "closed"), size = 1.5, linejoin = "round") +
  geom_segment(mapping = aes(x = 639055, y = -0.07,
                             xend = 638021, yend = -0.07),
               arrow = arrow(angle = 10, type = "closed"), size = 1.5, linejoin = "round") +
  labs(x = bquote(italic("3a") ~ "marker"),
       y = bquote("Variant relative frequency")) +
  annotation_custom(legendLabs[[1]]) +
  annotation_custom(legendLabs[[2]]) +
  annotation_custom(legendLabs[[3]]) +
  annotation_custom(legendLabs[[4]]) +
  annotation_custom(legendLabs[[5]]) +
  theme_bw() +
  theme(
        axis.ticks = element_line(size = 1.0, colour = "black"),
        axis.ticks.length = unit(0.25, "cm"),
        axis.text.x = element_text(size = 18, colour = "black", family = "Luxi Mono", angle = 90, vjust = 0.5),
        axis.text.y = element_text(size = 18, colour = "black", family = "Luxi Mono"),
        axis.title = element_text(size = 30, colour = "black"),
        legend.position = "none",
        panel.grid = element_blank(),
        panel.border = element_rect(size = 3.5, colour = "black"),
        panel.background = element_blank(),
        plot.margin = unit(c(0.3,1.2,0.0,0.3), "cm"),
        plot.title = element_text(hjust = 0.5, size = 30)) +
  ggtitle(bquote(bold(.(sample) ~ "ONT (" *
                      .(prettyNum(dim(plp2)[2]-4,
                                  big.mark = ",", trim = T)) *
                      " alignments)")))
ggsave(paste0(plotDir, sample, "_ONT_allele_relative_freq_v181219.pdf"),
       plot = ggObj_af,
       height = 6.5*1, width = 20, limitsize = F)
