#!/bin/bash

# author: Andy Tock
# date: 04/02/2022

# Apply mtsv2bedGraph_deepsignal.py (adapted to work with a DeepSignal raw read output TSV;
# original Nanopolish-based script mtsv2bedGraph.py from https://github.com/isaclee/nanopore-methylation-utilities )
# to generate a sorted, bgzipped, and tabix-indexed BED-style format methylation file
# Then, "Using the converted bed-style methylation file, the original bam file [e.g., generated by winnowmap]
# can be 'bisulfite converted in silico' for easy visualization on IGV via their bisulfite mode."
# Note that this latter step cannot be run with newer version of samtools (> v1.9) due to changes to the SAM format spec

# Usage on node7 (must be allocated to node10 as it requires > 300 GB RAM):
# csmit -m 400G -c 48 "bash ./per_read_tsv2bed_bam_conversion_CHH.sh Col_0_deepsignalDNAmeth_30kb_90pc_MappedOn t2t-col.20210610 48 5G CHH" 

prefix=$1
refbase=$2
threads=$3
sortMem=$4
context=$5

source activate ChIPseq_mapping

## Generate a sorted, bgzipped, and tabix-indexed BED-style format methylation file
## See https://github.com/isaclee/nanopore-methylation-utilities
## --offset may need to be changed from 0 to 1 for DeepSignal raw read output TSV,
## as coordinates are 0-based, whereas those for the equivalent Nanopolish output are 1-based
## I haven't done this in my tests with mtsv2bedGraph_deepsignal.py as it's not clear whether or not
## the BED-style output of mtsv2bedGraph.py should have 0-based or 1-based start coordinates
## (these should be 0-based if following BED format specs; additionally, mtsv2bedGraph.py adds 1
## to the final cytosine ("read end") coordinate, suggesting that start coordinates are 0-based and
## end coordinates are 1-based)
chrs=$(cut -f1 "/home/ajt200/analysis/nanopore/"${refbase}"/"${refbase}".fa.fai")
for chr in ${chrs};
do
  mtsv2bedGraph_deepsignal.py --input ${prefix}_${refbase}_${context}_raw_${chr}.tsv \
                              --mod ${context} \
                              --call-threshold 0.02 \
                              --genome /home/ajt200/analysis/nanopore/${refbase}/${refbase}.fa \
                              --offset 0 \
                              --verbose | \
    sort -k1,1 -k2,2n > ${prefix}_${refbase}_${context}_${chr}.bed
  cat ${prefix}_${refbase}_${context}_${chr}.bed
done | sort -k1,1 -k2,2n | bgzip > ${prefix}_${refbase}_${context}.bed.gz 
tabix -p bed ${prefix}_${refbase}_${context}.bed.gz
#${prefix}_${refbase}_${context}_${chr}.bed

# "Using the converted bed-style methylation file, the original bam file [e.g., generated by winnowmap]
# can be 'bisulfite converted in silico' for easy visualization on IGV via their bisulfite mode."
# See https://github.com/isaclee/nanopore-methylation-utilities
prefix_substr=$(echo ${prefix}_${refbase} | cut -d'_' -f 1,2,4-7)
context_lower=$(echo ${context} | awk '{print tolower($0)}')
convert_bam_for_methylation_deepsignal.py --threads ${threads} \
                                          --verbose \
                                          --fasta /home/ajt200/analysis/nanopore/${refbase}/${refbase}.fa \
                                          --bam ${prefix_substr}_winnowmap_sorted.bam \
                                          --mbed ${prefix}_${refbase}_${context}.bed.gz \
                                          --mod ${context_lower} | \
  /applications/samtools/samtools-1.9/samtools sort -@ ${threads} -m ${sortMem} -o ${prefix}_${refbase}_${context}_winnowmap_sorted_BSconverted_sorted.bam
/applications/samtools/samtools-1.9/samtools index ${prefix}_${refbase}_${context}_winnowmap_sorted_BSconverted_sorted.bam

conda deactivate
